File: libs/jgdsl/src/gdsl/asm/operand/Sum.java
package gdsl.asm.operand;

import gdsl.asm.Visitor;

public class Sum extends Operand {
  private Operand lhs;
  
  public Operand getLhs() {
    return lhs;
  }
  
  private Operand rhs;
  
  public Operand getRhs() {
    return rhs;
  }

  /**
   * @param lhs
   * @param rhs
   */
  public Sum (Operand lhs, Operand rhs) {
    super();
    this.lhs = lhs;
    this.rhs = rhs;
  }

  @Override public String toString () {
    return "(" + lhs + " + " + rhs + ")";
  }

  @Override public void accept (Visitor v) {
    v.visit(this);
  }
}

File: libs/cppgdsl/src/assembly/operand/sum.cpp
#include "cppgdsl/assembly/operand/sum.h"
#include "cppgdsl/assembly/operand/operand_visitor.h"

namespace gdsl {
namespace assembly {

void sum::put(std::ostream& out) const {
  out << "(" << *lhs << " + " << *rhs << ")";
}

void sum::accept(operand_visitor& ov) const { ov.visit(this); }

bool sum::operator==(operand const& o) const {
  bool equals = false;
  operand_visitor v(true);
  v._([&](sum const* o) {
    equals = *this->lhs == *o->lhs && *this->rhs == *o->rhs;
  });
  o.accept(v);
  return equals;
}

}  // namespace assembly
}  // namespace gdsl
